src/engine/TPA.cc

```diff
@@ -1691,6 +1694,40 @@ InvalidityWitness TransitionSystemNetworkManager::computeInvalidityWitness() con
     return InvalidityWitness::fromErrorPath(path, graph);
 }
 
+ValidityWitness TransitionSystemNetworkManager::computeValidityWitness() {
+    if (not isTransitionSystemDAG(graph)) { return {}; }
+    auto vertices = graph.getVertices();
+    assert(vertices.size() >= 3);
+    TermUtils utils(logic);
+    TimeMachine timeMachine(logic);
+    TermUtils::substitutions_map subs;
+    std::vector<ValidityWitness> witnesses;
+    ValidityWitness::definitions_t definitions;
+
+    for (auto node : vertices) {
+        auto vertex = node;
+        if (vertex == graph.getEntry() || vertex == graph.getExit()) continue;
+
+        auto graphVars = utils.predicateArgsInOrder(graph.getStateVersion(vertex));
+        vec<PTRef> unversionedVars;
+        auto systemVars = networkMap.at(vertex).solver->getStateVars(0);
+
+        for (std::size_t i = 0; i < graphVars.size(); ++i) {
+            unversionedVars.push(timeMachine.getUnversioned(graphVars[i]));
+            subs.insert({systemVars[i], unversionedVars.last()});
+        }
+        getNode(vertex).solver->resetInitialStates(getNode(vertex).trulySafe);
+        auto [res, explanation] = queryTransitionSystem(networkMap.at(vertex));
+        assert(res == ReachabilityResult::UNREACHABLE);
+        if (res == ReachabilityResult::UNREACHABLE) {
```

Comment:
I think we should add an `else` case where we return `NoWitness` and explain what went wrong.

Reply:
To be honest, is should always be true, if it is REACHABLE, it means there is a problem in the solving algorithm(as it wasn't able to cover all passes)

The reason I've even added if, is that CI throws an error that `res` is not used otherwise

Reply:
I understand that it SHOULD be true :)
I agree that if it is not, something has gone wrong, so we should not pretend that we have the witness, and instead be safe and return no witness.
It can be something like `return NoWitness("Unexpected situation occurred during witness computation in TPA engine");`

Reply:
I agree with general idea, I just don't like the idea of `computeValidityWitness` returning `variant<NoWitness, ValidityWitness>` 😅

Reply:
I see, OK. How about immediately returning empty `ValidityWitness`?